extra-tests/classes/RollupStateTests.cls

```diff
@@ -0,0 +1,371 @@
+    for (Integer index = 0; index < 400; index++) {
+      Rollup__mdt clonedMeta = template.clone();
+      clonedMeta.DeveloperName += '' + index;
+      RollupState.GenericInfo info = (RollupState.GenericInfo) state.getState(stubAccountId, clonedMeta, RollupState.GenericInfo.class);
+      info.value = index;
+      if (index < 20) {
+        statesToSample.add(info.getUntypedState());
+      }
+    }
+    state.commitState(stubJobId);
+
+    List<RollupState__c> committedStates = [
+      SELECT Body0__c, RelatedRecordKeys0__c
+      FROM RollupState__c
+    ];
+
+    Assert.areEqual(3, committedStates.size());
+    Boolean body0Filled = false;
+    Boolean sampleStatesFilled = false;
+    for (Integer index = 0; index < committedStates.size(); index++) {
+      RollupState__c committedState = committedStates[index];
+      body0Filled = body0Filled || committedState.Body0__c != null;
+      Assert.areEqual(stubAccountId, committedState.RelatedRecordKeys0__c, 'State was missing key at index: ' + index);
+      if (sampleStatesFilled) {
+        Boolean isDuplicate = committedState.Body0__c.contains(JSON.serialize(statesToSample).removeStart('[').removeEnd(']'));
+        if (isDuplicate) {
+          throw new IllegalArgumentException('Body0__c should not match for both states');
+        }
+      } else {
+        sampleStatesFilled = committedState.Body0__c?.contains(JSON.serialize(statesToSample).removeStart('[').removeEnd(']')) == true;
+      }
+    }
+
+    Assert.isTrue(body0Filled);
+    Assert.isTrue(sampleStatesFilled, committedStates[0].Body0__c?.substring(0, 50));
+    /**
+     * ensure that we don't exceed the DataWeave heap:
+     * System.DataWeaveScriptException: turtles.api.SandboxedLimitException - Heap limit exceeded 6004410 > 6000000
+     */
+    state.loadState(stubJobId, new Set<String>{ stubAccountId });
+  }
+
+  @IsTest
+  static void splitsRelatedRecordKeysIntoDifferentTextFields() {
+    Rollup__mdt template = new Rollup__mdt(DeveloperName = 'one rollup');
+    RollupState state = new RollupState();
+    Set<String> fullKeys = new Set<String>();
+
+    List<String> fieldKeys0 = new List<String>();
+    List<String> fieldKeys1 = new List<String>();
+    List<String> fieldKeys2 = new List<String>();
+    List<String> fieldKeys3 = new List<String>();
+    List<String> fieldKeys4 = new List<String>();
+    List<String> fieldKeys5 = new List<String>();
+    List<String> fieldKeys6 = new List<String>();
+    List<String> fieldKeys7 = new List<String>();
+    List<String> fieldKeys8 = new List<String>();
+    List<String> fieldKeys9 = new List<String>();
+    List<String> fieldKeys10 = new List<String>();
+    List<String> secondFieldKeys0 = new List<String>();
+    // 255 / 18 characters ~= 13 records per field (accounting for commas)
+    String firstAccountId = RollupTestUtils.createId(Account.SObjectType) + 'aaa';
+    for (Integer index = 0; index < 17 * 15; index++) {
```

Comment:
2 questions here:

1. Where do `17` and `15` come from here? 
    - I was thinking `17` should be `18` for the 18 character ID length?
2. Do you think it's worth adding another similar test, where the keys are `255` length (the max possible for an external ID field), just to make sure things still work with that length/scenario?

Reply:
I made some explanatory comments on this test to try to sew it all up! As for question 2, there _is_ another test where I override the static variable (`RollupStateTests.alwaysTracksRelatedRecordKeys()`) to only allow a single key per field, so I think that particular scenario is already covered